### PR TITLE
Split multi output

### PR DESCRIPTION
**Pull request by angeloskath**
Multi output primitive implementing splitting across an axis. It only routes to the primitive when the splits are all non-empty and contiguous. This allows to use concatenate to gather back all the gradients which is faster than scatter_add.

Using scatter_add is necessary when having overlapping slices so this case still uses slices. Also we use slices when performing more than 10 splits just to be safe regarding multi-output primitives.

The following benchmark

```python
def qkvfn(x):
    q, k, v = [xi.squeeze(0) for xi in x.split(3)]
    return (mx.softmax(q @ k.swapaxes(-2, -1)) @ v).sum()

def grad_fn(x):
    l, dfdx = mx.value_and_grad(qkvfn)(x)
    mx.eval(l, dfdx)
    
x = mx.random.normal(shape=(3, 256, 128, 256))
time_fn(grad_fn, x)
```

goes from **39ms** to **14ms** on my M2 air.

### CODE REVIEW

**Comment by awni** [comment]
Should we update the iterator version as well? I feel like it's quite common to use that syntax rather than an explicit split.

```
a, b, c = arr
```

The [ArrayIterator](https://github.com/ml-explore/mlx/blob/main/mlx/array.h#L124) would probably need a `vector<array>` to store the splits. Could do something similar there: if the splits < 10, use split and populate the vector and retrieve from it, otherwise use the slice as it is now. Wdyt?

**Comment by awni** [comment]
Do you mind adding a test for the grad / vjp of the multi-output split? Your implementation looks correct and simple, but it's the first real use case of mult-output primitives for that (which could still have some lurking issues).

File: mlx/primitives.cpp [hunk 1]
```diff
@@ -2488,6 +2488,32 @@ bool Sort::is_equivalent(const Primitive& other) const {
...
+    const std::vector<int>& axes) {
+  return {
+      {split(inputs[0], indices_, axis_ + (axes[0] <= axis_), stream())}, axes};
+}
+
+std::vector<array> Split::vjp(
+    const std::vector<array>& primals,
+    const std::vector<array>& cotangents,
+    const std::vector<int>& argnums) {
+  return {concatenate(cotangents, axis_, stream())};
```
Comment (awni) [inline mlx/primitives.cpp:2502]: Beautiful

File: mlx/ops.cpp [hunk 2]
```diff
@@ -573,6 +574,29 @@ std::vector<array> split(
...
+      indices[0] > 0 && indices.back() < a.shape(ax)) {
+    std::vector<Dtype> dtypes(indices.size() + 1, a.dtype());
+    std::vector<std::vector<int>> shapes(indices.size() + 1, a.shape());
+    shapes[0][ax] = indices[0];
+    for (int i = 1; i < indices.size(); i++) {
+      shapes[i][ax] = indices[i] - indices[i - 1];
+    }
+    shapes.back()[ax] = a.shape(ax) - indices.back();
+
+    return array::make_arrays(
```
Comment (awni) [inline mlx/ops.cpp:593]: Not too bad to use a multi-output primitive eh 😉

**Comment by awni** [comment]
This is really nice!! And I'm glad to see the multi-output coming in handy after all that work, I love that benchmark!

**Comment by angeloskath** [comment]
> Thanks for adding that! We can merge whenever, but I think it would be good to add a test for the split VJP (since it's our first multi-output VJP)

Yep that's why I didn't merge yesterday ;-) .

**Comment by angeloskath** [comment]
@awni let me know if you want a more extensive test. I think this covers the main use-case quite well and checks for proper placement of `zeros` at "unknown" cotangents.

**Comment by awni** [comment]
That's perfect, thank you 🙏 !

**Pull request merged**